src/pykeen/nn/node_piece.py

```diff
@@ -0,0 +1,667 @@
+
+        :param anchors: shape: (a,)
+            the anchor node IDs
+        :param adjacency: shape: (n, n)
+            the adjacency matrix
+        :param max_iter:
+            the maximum number of hops to consider
+        :param k:
+            the minimum number of anchor nodes to reach
+
+        :return: shape: (n, a)
+            a boolean array indicating whether anchor $j$ is in the set of $k$ closest anchors for node $i$
+        """
+        num_entities = adjacency.shape[0]
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        # an array storing whether node i is reachable by anchor j
+        reachable = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable[anchors] = numpy.eye(num_anchors, dtype=bool)
+
+        # an array indicating whether a node is closed, i.e., has found at least $k$ anchors
+        final = numpy.zeros(shape=(num_entities,), dtype=bool)
+
+        # the output
+        pool = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+
+        old_reachable = reachable
+        for i in range(max_iter):
+            # propagate one hop
+            reachable = adjacency.dot(reachable)
+            # convergence check
+            if (reachable == old_reachable).all():
+                logger.warning(f"Search converged after iteration {i} without all nodes being reachable.")
+                break
+            old_reachable = reachable
+            # copy pool if we have seen enough anchors and have not yet stopped
+            num_reachable = reachable.sum(axis=1)
+            enough = num_reachable >= k
+            mask = enough & ~final
+            logger.debug(
+                f"Iteration {i}: {format_relative_comparison(enough.sum(), total=num_entities)} closed nodes.",
+            )
+            pool[mask] = reachable[mask]
+            # stop once we have enough
+            final |= enough
+            if final.all():
+                break
+        return pool
+
+    @staticmethod
+    def select(
+        pool: numpy.ndarray,
+        k: int,
+    ) -> numpy.ndarray:
+        """
+        Select $k$ anchors from the given pools.
+
+        :param pool: shape: (n, a)
+            the anchor candidates for each node (a binary array)
+        :param k:
+            the number of candidates to select
+
+        :return: shape: (n, k)
+            the selected anchors. May contain -1 if there is an insufficient number of  candidates
+        """
+        tokens = numpy.full(shape=(pool.shape[0], k), fill_value=-1, dtype=int)
+        generator = numpy.random.default_rng()
+        # TODO: can we replace this loop with something vectorized?
+        for i, row in enumerate(pool):
+            (this_pool,) = row.nonzero()
+            chosen = generator.choice(a=this_pool, size=min(k, this_pool.size), replace=False, shuffle=False)
+            tokens[i, : len(chosen)] = chosen
+        return tokens
```

Comment:
@migalkin 

this loop seems to be reasonably fast ( ~13sec for YAGO3-10 with 10_000 anchors, and 20 selected).

Reply:
Confirm, this one is pretty snappy (even less than 13 sec on my laptop) 🚀